QUIC Demuxer and Record Layer (RX+TX)

hlandau opened the pull request:
Here we go.

RX only for now aside from some small currently-vestigial pieces of TX functionality.

Key update is mandatory AIUI but involves interaction between the TX/RX pieces so the plan is:
1. Develop the TX side on top of this PR and update this PR
2. Add tests which get the TX and RX sides talking to each other
3. Key update and key update tests

So this shouldn't be merged yet (well, it could be, and I just make a new PR), but reviews are welcome.

This includes a demuxer because implementing the QRL resulted in 90% of the functionality of a demuxer and it made no sense not to include it, since there's necessary interaction between the two.

mattcaswell commented on ssl/quic/quic_record_util.c:
```diff
@@ -0,0 +1,194 @@
+    return ret;
+}
+
+int ossl_quic_hkdf_expand_label(OSSL_LIB_CTX *libctx,
```
Looks very similar to `tls13_hkdf_expand`

hlandau replied:
(see above)

mattcaswell commented on include/internal/quic_types.h:
```diff
@@ -11,6 +11,37 @@
+        case QUIC_ENC_LEVEL_HANDSHAKE:
+            return QUIC_PN_SPACE_HANDSHAKE;
+        default:
+            assert(0);
```
Strange having the default case in the middle.

mattcaswell commented on include/internal/quic_demux.h:
```diff
@@ -0,0 +1,251 @@
+typedef struct quic_urxe_st QUIC_URXE;
+
+/* Maximum number of packets we allow to exist in one datagram. */
+#define QUIC_MAX_PKT_PER_URXE       (sizeof(uint64_t) * 8)
```
I wonder why we need this? Is this restriction in the RFC? Is it likely that we might get more than that?

hlandau replied:
So, the underlying issue is that we should process each packet only once. i.e., we do not want a case where the QRL returns the same packet twice. This could otherwise happen when we receive a datagram with packets belonging to several different ELs, some of which we don't have keys for yet. This is common. So when we process a datagram, we might decrypt and queue some of those packets and defer others because we don't have a key for them yet. These datagrams are reprocessed when we call `ossl_qrl_provide_rx_secret` for the relevant EL. But we don't want to process and return the packets we already processed again, thus we need a way to record whether some (but not all) packets in a given datagram have already been returned.

Originally I was going to use a complex algorithm which reused the memory of processed packets in a datagram to allow packets to be recorded as deleted, but this was far more trouble than it's worth. The major purpose of allowing multiple packets in a datagram is to allow multiple ELs to be sent in one datagram. There are only four ELs. 64 packets per datagram is more than should be ever needed by quite a wide margin.

mattcaswell replied:
Hmm. Yeah, ok.

mattcaswell commented on include/internal/quic_types.h:
```diff
@@ -26,4 +57,19 @@ static ossl_unused ossl_inline QUIC_PN ossl_quic_pn_min(QUIC_PN a, QUIC_PN b)
+{
+    if (a->id_len != b->id_len || a->id_len > QUIC_MAX_CONN_ID_LEN)
+        return 0;
+    return !memcmp(a->id, b->id, a->id_len);
```
Should do `memcmp() == 0` instead

mattcaswell commented on include/internal/quic_wire_pkt.h:
```diff
@@ -0,0 +1,435 @@
+ *   needed. If not enough sample bytes are provided, this function fails.
+ *
+ * first_byte:
+ *   The initial byte of the QUIC packet to be decrypted.
```
Does this mean the initial byte of the QUIC packet header? (as opposed to the initial byte of the QUIC packet payload)

hlandau replied:
Yes. I'll tweak the text.

mattcaswell reviewed (changes requested):
Mostly just looked at the headers so far.

hlandau commented on include/internal/quic_demux.h:
```diff
@@ -0,0 +1,251 @@
+};
+
+/* Accessors for URXE buffer. */
+static ossl_unused ossl_inline uint8_t *
```
Force of habit... will fiix.

hlandau commented on include/internal/quic_demux.h:
```diff
@@ -0,0 +1,251 @@
+/*
+ * Unregisters any datagram handler callback from all connection IDs it is used
+ * for. cb and cb_arg must both match the values passed to
+ * ossl_quic_demux_register.
```
The latter. The normal way to unregister is via `ossl_quic_demux_unregister`, this just provides a convenient way for a QRL to unregister every DCID it's registered with the demuxer when it's freed. Otherwise it would have to track this information itself.

hlandau commented on include/internal/quic_record.h:
```diff
@@ -0,0 +1,311 @@
+ * A QRL has several encryption levels (Initial, Handshake, 0-RTT, 1-RTT) and
+ * two directions (RX, TX). At any given time, key material is managed for each
+ * (EL, RX/TX) combination.
+ *
```
It depends on how closely TX and RX end up being coupled due to the requirements of QUIC. I intend to implement TX in the same object for now, but if having done that they seem cleanly separable I would probably then separate them.

hlandau commented on include/internal/quic_record.h:
```diff
@@ -0,0 +1,311 @@
+ *
+ * To transition the RX side of an EL from WAITING_FOR_KEYS to HAVE_KEYS, call
+ * ossl_qrl_provide_rx_secret (or for the INITIAL EL,
+ * ossl_qrl_provide_rx_secret_initial).
```
I assume if we did that we would also want a polymorphic interface?

I've looked at the existing KTLS socket interface and I don't think supporting it would be too much trouble with the current design, if someday KTLS is extended to support QUIC. The demuxer owns the RX BIO so it would need to configure it for the RX side, and it can have a flag in the URXE it passes indicating if a packet is already decrypted (which could vary by encryption level). Then we just skip decryption in the QRL if the flag is set for a URXE. Probably the call to `ossl_qrl_provide_rx_secret` would pass the secret along to the demuxer.

Basically I don't think adding KTLS support to this when KTLS-for-QUIC happens would be much of a maintenance burden. Having an entirely separate implementation of the QRL for KTLS seems overkill to me and would duplicate most of the code.

mattcaswell replied:
In TLS I handle this by separating the common code out. So, in #18132, I have tls_common.c for the all the common code paths, and then tls13_meth.c for TLSv1.3 specific stuff, ktls_meth.c for KTLS specific stuff, etc. I have separate `OSSL_RECORD_METHOD` objects for TLSv1.3 (standard OpenSSL implementation), KTLS, DTLS, etc. Each `OSSL_RECORD_METHOD` uses some code from tls_common.c and some code from the record method specific file.

hlandau commented on include/internal/quic_record.h:
```diff
@@ -0,0 +1,311 @@
+ *
+ *   - In the DISCARDED state, the QRL assumes no keys for the given
+ *     EL will ever arrive again. If it receives any packet for an EL
+ *     in this state, it is simply discarded.
```
There is no 'current' EL (as far as the QRL is concerned), each EL has its own state of whether it has keys or not. So you can receive on all ELs simultaneously so long as you've provided keys for all of them.

The only caveat to this is that the QRL does auto-discard INITIAL keys when it first receives a HANDSHAKE packet (which is a requirement straight from the RFC).

mattcaswell replied:
Ok. Makes sense.

hlandau commented on include/internal/quic_record_util.h:
```diff
@@ -0,0 +1,79 @@
+                          unsigned char *out, size_t out_len);
+
+/* HKDF-Expand-Label(Secret, Label, Context, Length) (RFC 8446 - TLS 1.3) */
+int ossl_quic_hkdf_expand_label(OSSL_LIB_CTX *libctx,
```
Almost &mdash; the problem is that `tls13_hkdf_expand` depends on a `SSL_CONNECTION` object.

Maybe it could be refactored? The only other concern I have is use of `ERR_raise`. Since this function can be triggered from unauthenticated network traffic. (Basically, since in QUIC, key update is triggered by the key phase bit in the header, an attacker can initiate key update attempts by flipping the key phase bit randomly. This will not ultimately succeed but we do have to do some key derivation first to determine whether the key update is legitimate or not.) Although those are only internal/allocation errors, so I guess nothing to really be concerned about.

`tls13_hkdf_expand` could be refactored to call another function with `sctx->libctx` and `sctx->propq`, allowing QRL to also use it? Works for me.

mattcaswell replied:
Refactoring it makes sense to me. Having 2 implementations of essentially the same thing seems like a bad idea to me. We should reuse where possible/practical.

hlandau commented on ssl/quic/quic_record_util.c:
```diff
@@ -0,0 +1,194 @@
+    0x74, 0x6C, 0x73, 0x31, 0x33, 0x20 /* "tls13 " */
+};
+
+int ossl_quic_hkdf_extract(OSSL_LIB_CTX *libctx,
```
`tls13_generate_secret` uses a fixed key and has a few differences. I could look at refactoring the two to share the common parts?

mattcaswell replied:
Refactoring would be a good idea.

hlandau commented:
Not sure what's going on with a/ub/msan. I can't replicate it on my machine with either clang or GCC sanitizers, and the error reported isn't even a sanitizer error... the test is supposed to be deterministic. Hm...

hlandau commented:
Just some nits for now.

hlandau commented:
I've looked at the use of OSSL_RECORD_METHOD by TLS and KTLS... IMO it's going to be premature to do something like this for now, especially since the KTLS interface for QUIC literally doesn't exist yet. It makes more sense when needing to support four versions of TLS, DTLS plus KTLS, but the scope of variation for QUIC is a lot narrower. I think this can wait for now; moreover, it's either a question of refactoring this now with an unknown design space (future KTLS APIs) which will probably have to be redone later, or later when we actually know what's needed. So, I'd like to avoid introducing polymorphism into the QRL for now.

The msan issue is weird. Going to have to try and repro it on a build server.

mattcaswell commented:
> I've looked at the use of OSSL_RECORD_METHOD by TLS and KTLS... IMO it's going to be premature to do something like this for now, especially since the KTLS interface for QUIC literally doesn't exist yet. It makes more sense when needing to support four versions of TLS, DTLS plus KTLS, but the scope of variation for QUIC is a lot narrower. I think this can wait for now; moreover, it's either a question of refactoring this now with an unknown design space (future KTLS APIs) which will probably have to be redone later, or later when we actually know what's needed. So, I'd like to avoid introducing polymorphism into the QRL for now.

Fair enough.

hlandau commented:
The test issue was caused by random test ordering. Fixed.

hlandau commented:
Updated, refactored.

The other KDF function is similar to `tls13_generate_secret` but not a match. It doesn't use the same label found in that function, and uses HKDF instead of `TLS1_3_KDF`. So I've kept this separate.

I think this settles the outstanding feedback.

hlandau commented:
Updated:
- Adds TX side.
- The QRL has been split into independent RX (QRX) and TX (QTX) objects as I'm now satisfied they can be separated reasonably cleanly.
- QRL (the RX side) has been renamed to QRX.
- Some tests for the TX side.
- Key expiration functionality has been added.
- A preview of the API and logical model for key update can be viewed in the header files `quic_record_rx.h` and `quic_record_tx.h`. No implementation yet.

hlandau commented:
Also while I remember: @paulidale, let me know if it would be more helpful for the iovecs to be a linked list instead of an array.

levitte commented:
Do I understand correctly that this replaces #18870?

levitte commented on include/internal/quic_record_rx.h:
```diff
@@ -0,0 +1,488 @@
+     * datagrams containing INITIAL packets), as required by RFC 9000.
+     */
+    size_t              datagram_len;
+} OSSL_QRX_PKT;
```
As said [here](https://github.com/openssl/openssl/pull/18870#issuecomment-1195122350), I'd like to get the reception time of the packet as well.  At the time, I was thinking that it could be done by filling in an `OSSL_TIME` pointer argument to the `ossl_qrx_read_pkt()` call, but ideally, it should be an `OSSL_TIME` field in this structure.

hlandau commented:
> Do I understand correctly that this replaces #18870?

Yeah, it does.

levitte commented on include/internal/quic_record_rx.h:
```diff
@@ -0,0 +1,488 @@
+ * On success, all fields of *pkt are filled and 1 is returned.
+ * Else, returns 0.
+ *
+ * The resources referenced by pkt->hdr, pkt->data and pkt->peer will remain
```
.... except, there's no `data` field!

levitte replied:
Also, there needs to be a `data_len` field

levitte replied:
Oh, found them in `QUIC_PKT_HDR`...  which is, quite frankly, a bit surprising, but ok.

levitte commented on ssl/quic/quic_wire_pkt.c:
```diff
@@ -0,0 +1,734 @@
+            if (hdr->type == QUIC_PKT_TYPE_RETRY) {
+                /* Retry packet. */
+                hdr->data       = PACKET_data(pkt);
+                hdr->len        = PACKET_remaining(pkt);
```
Uhmmm, unless I'm reading https://datatracker.ietf.org/doc/html/rfc9000#section-17.2.5 wrong, the RETRY packet has no payload.  It does, however, have a token (which should end up in `hdr->token`) and a token integrity tag.

hlandau replied:
Yes, the RFC calls it a retry token plus retry integrity instead of a payload. In actual functional terms this is the payload of the packet, so my packet decoder takes the pragmatic view of just considering this to be the payload of a Retry packet. I don't think it makes sense to have separate fields for this in the header structure.

levitte replied:
Point is that you already have the `token` field...  that's where I'd expect to find the retry token!

hlandau replied:
That's found in the Initial packet header and is something very different... the Retry Token resembles a payload much more than it resembles the Initial Token.

levitte replied:
Ok, fair enough

hlandau commented:
Updated:

- Adds an RX time field to the OSSL_QRX_PKT structure.
    
- Adds a timekeeping argument to ossl_demux_new which is used to determine
      packet reception time.
    
- Adds a decoded PN field to the OSSL_QRX_PKT structure.
      This has to be decoded by the QRX anyway, and its omission was an oversight.
    
- Key update support for the TX side.
    
- Minor refactoring.

levitte commented on include/internal/quic_record_rx.h:
```diff
@@ -0,0 +1,507 @@
+ * ossl_qrx_read_pkt. Pass the opaque value pkt->handle returned in the
+ * structure.
+ */
+void ossl_qrx_release_pkt(OSSL_QRX *qrx, void *handle);
```
Is there a reason why the opaque handle must be passed?  Can't we just as well pass the `OSSL_QRX_PKT` pointer and leave it to `ossl_qrx_release_pkt()` to use the proper bits?

hlandau replied:
We could. I guess the issue there is that it requires the caller to keep `OSSL_QRX_PKT` around until it wants to release the resources. That isn't really the intention, since `OSSL_QRX_PKT` is intended as just a temporary structure for returning information. So you could throw away the structure immediately after calling `ossl_qrx_read_pkt` and store the information in it in another way, so long as you keep the handle.

levitte replied:
Ok, there seems to be different ideas about this.  See the discussion starting here: https://github.com/openssl/openssl/pull/18838#discussion_r940383537

hlandau replied:
Responded over there

paulidale commented on include/internal/quic_types.h:
```diff
@@ -11,6 +11,38 @@
+ossl_quic_enc_level_to_pn_space(uint32_t enc_level)
+{
+    switch (enc_level) {
+        case QUIC_ENC_LEVEL_INITIAL:
```
Indentation nit.

t8m commented:
Can you please rebase this?

hlandau commented:
@t8m Done.

levitte commented:
There's a thing I didn't notice before: `ossl_quic_wire_encode_padding()` and `ossl_quic_wire_decode_padding()`, is there a reason they don't have `_frame_` in the name?

hlandau commented:
I mean, they encode/decode multiple frames. I think there was a previous comment on the naming in the PR that added them.

t8m commented on include/internal/quic_record_rx.h:
```diff
@@ -0,0 +1,507 @@
+ *      Two keys and a timer
+ *
+ *      "Alternatively, endpoints can retain only two sets of packet protection
+ *       neys, swapping previous keys for next after enough time has passed to
```
Typo: keys

t8m commented on include/internal/quic_record_rx.h:
Should this be handled via a callback? Otherwise you'd have to always call it after every ossl_qrx_read_pkt() call. Or, alternatively have it as an output parameter to ossl_qrx_read_pkt()?

hlandau replied:
Seems reasonable. I'll add it.

t8m commented on include/internal/quic_record_rx.h:
There are two limits - confidentiality limit - that should limit the number of packets sent with a given key. This is TX side limit so should be handled on the TX record layer.

The integrity limit (i.e., the count of received packets that weren't successfully decrypted and authenticated) applies to the RX side and if I understand the RFC 9001 correctly it applies across all keys in the connection so it is not per epoch. There is also no way how to escape the integrity limit by key update - if the limit is reached `the endpoint MUST immediately close the connection with a connection error of type AEAD_LIMIT_REACHED and not process any more packets.`

t8m commented on include/internal/quic_record_rx.h:
```diff
@@ -0,0 +1,507 @@
+
+/*
+ * Provides a secret to the QRX, which arises due to an encryption level change.
+ * enc_level is a QUIC_ENC_LEVEL_* value. To initialise the INITIAL encryption
```
I do not see QUIC_ENC_LEVEL being defined anywhere.

hlandau replied:
It's defined in `include/internal/quic_types.h`

t8m commented on include/internal/quic_record_tx.h:
```diff
@@ -0,0 +1,299 @@
+
+/*
+ * Provides a secret to the QTX, which arises due to an encryption level change.
+ * enc_level is a QUIC_ENC_LEVEL_* value.
```
Same as in rx side for QUIC_ENC_LEVEL

t8m commented on include/internal/quic_record_tx.h:
```diff
@@ -0,0 +1,299 @@
+    QUIC_PKT_HDR               *hdr;
+
+    /*
+     * iovecs expressing the logical packet payload buffer. Zero-length entries
+     * are permitted.
```
Why do we need to allow zero-length entries?

t8m commented on include/internal/quic_record_tx.h:
```diff
@@ -0,0 +1,299 @@
+ * It is the caller's responsibility to ensure it uses a consistent DCID length
+ * for communication with any given set of remote peers.
+ *
+ * The packet is queued regardless of whether it is able to be sent immediately.
```
Perhaps add some comment why? I.E., to support use of the sendmmsg().

hlandau replied:
Will do.

t8m reviewed (changes requested):
Good work! Review in progress...

levitte commented on include/internal/quic_record_rx.h:
```diff
@@ -0,0 +1,507 @@
+# include "internal/quic_demux.h"
+
+/*
+ * QUIC Record Layer - RX
+ * ======================
+ */
+typedef struct ossl_qrx_st OSSL_QRX;
```
> By the way, the OSSL_ prefix is not needed on structures

Do we think we will ever make these names public?  That needs to be factored in.

t8m replied:
IMO no. And we could/should rename them appropriately if we decide so.

hlandau commented on include/internal/quic_record_rx.h:
```diff
@@ -0,0 +1,507 @@
+ *
+ * To transition the RX side of an EL from WAITING_FOR_KEYS to HAVE_KEYS, call
+ * ossl_qrx_provide_secret (for the INITIAL EL, use of
+ * ossl_qrl_provide_initial_secret is recommended).
```
Fixed, it's ossl_quic_provide_initial_secret.

hlandau commented:
Updated.

I'm a bit wary of changing the name. Some identifiers are already quite long. But I'll take suggestions.

levitte commented on ssl/quic/quic_demux.c:
```diff
@@ -0,0 +1,535 @@
+void ossl_quic_urxe_remove(QUIC_URXE_LIST *l, QUIC_URXE *e)
+{
+    /* Must be in list currently. */
+    OPENSSL_assert((e->prev != NULL || l->head == e)
```
This is preferred:

``` C
    if (!ossl_assert(expr)) {
        /* Do error recovery / return */
    }
```

https://github.com/openssl/openssl/blob/200d9521a0d406a7d02778d1c6c5a5230caeecf5/include/internal/common.h#L20-L35

hlandau commented on include/internal/quic_record_rx.h:
```diff
@@ -489,13 +500,16 @@ int ossl_qrx_key_update_timeout(OSSL_QRX *qrx, int normal);
  */
-uint64_t ossl_qrx_get_cur_epoch_forged_pkt_count(OSSL_QRX *qrx,
-                                                 uint32_t enc_level);
+uint64_t ossl_qrx_get_cur_epoch_forged_pkt_count(OSSL_QRX *qrx);
```
Yeah, just noticed that...

hlandau commented on include/internal/quic_wire_pkt.h:
```diff
@@ -0,0 +1,519 @@
+    unsigned int    partial     :1;
+
+    /*
+     * [ALL] Whether the fixed bit was set. Note that only Version Negotiation
+     * packets are allowed to have this unset, so this will always be 1 for all
+     * other packet types (decode will fail if it is not set). Ignored when
+     * encoding unless encoding a Version Negotiation packet.
+     */
+    unsigned int    fixed       :1;
```
It lets us choose the value when encoding. It also becomes relevant with future RFCs like "Greasing the QUIC bit" where the fixed bit could change, though this only supports it for Version Negotiation packets currently.

hlandau commented:
Updated, rebased.

t8m commented on ssl/quic/quic_record_rx.c:
```diff
@@ -0,0 +1,1160 @@
+static void qrx_cleanup_rxl(RXE_LIST *l)
+{
+    RXE *e, *enext;
+    for (e = l->head; e != NULL; e = enext) {
```
Nit: empty line missing

t8m commented on ssl/quic/quic_record_rx.c:
```diff
@@ -0,0 +1,1160 @@
+static void qrx_cleanup_urxl(OSSL_QRX *qrx, QUIC_URXE_LIST *l)
+{
+    QUIC_URXE *e, *enext;
+    for (e = l->head; e != NULL; e = enext) {
```
Nit: empty line missing

t8m commented on ssl/quic/quic_record_rx.c:
```diff
@@ -0,0 +1,1160 @@
+     */
+    rxe2 = OPENSSL_realloc(rxe, sizeof(RXE) + n);
+    if (rxe2 == NULL)
+        /* original RXE is still in tact unchanged */
```
Typo: intact or in fact?

t8m commented on ssl/quic/quic_record_rx.c:
Nit: indentation is off by 1 space

hlandau commented:
Updated (nits).

t8m reviewed (changes requested):
Nits mostly. Great work!

hlandau commented on ssl/quic/quic_record_rx.c:
Not quite. We skip packets we already processed, but we do need to parse the headers minimally again to get the length so we can skip over it.

t8m replied:
Then I still do not understand the condition as currently if already processed the packet then we `goto malformed`?

hlandau replied:
If we have already processed the packet we set the flag `already_processed` so that we skip processing of the packet *after* we have discerned its length, which is the minimum we need to be able to skip over it.

We set `eop` to point to the end of the current packet before the above line. The label `malformed:` handles things accordingly. If we have eop as non-NULL we skip to the next packet as we were able to discern the packet length. Otherwise we have no idea if another packet follows or not and just abandon processing.

I guess the label `malformed` could be considered misleading? It's really `malformed_or_already_processed`.

t8m replied:
Would it make sense to instead do here:
```
if (already_processed) {
    ignore_res(PACKET_forward(pkt, eop - PACKET_data(pkt)));
    return 0;
}
```
Or at least add a comment like `/* We treat already_processed packets in dgram like malformed packets, i.e., ignoring them. */`

hlandau replied:
Added comment. I don't think it's worth optimising out the call to pkt_mark since it just sets a bit.

t8m replied:
I do not see the comment being pushed.

t8m replied:
Hmm, now I see it, I probably did not reload the GH view.

hlandau commented on ssl/quic/quic_record_rx.c:
```diff
@@ -0,0 +1,1160 @@
+        && rxe->hdr.version != QUIC_VERSION_NONE)
+        return 0;
+
+    /* Clients should never receive 0-RTT packets. */
+    if (rxe->hdr.type == QUIC_PKT_TYPE_0RTT)
```
Yeah, seems reasonable. I'll get on this at some point.

t8m replied:
`/* TODO(QUIC): ... */` would WFM too.

hlandau commented:
Updated, minor changes.

hlandau commented:
Added comment.

paulidale commented on include/internal/quic_demux.h:
```diff
@@ -0,0 +1,265 @@
+/*
+ * Creates a new demuxer. The given BIO is used to receive datagrams from the
+ * network using BIO_recvmmsg. short_conn_id_len is the length of destination
+ * connection IDs used in RX'd packets; it must have the same value for all
```
For a single connection, the IDs have to be consistently sized.  But across multiple connections, they don't and won't be.  According to the design, the demuxer will be shared between multiple connection objects.  The design receives all packets on one socket and then dispatches them appropriately.  In order to do this, the source host and port are required to determine the connection ID size.

For MVP this isn't required but it would be prudent to consider it now.

t8m replied:
These are destination connection IDs on a server side for short header packets. We could actually make the short_conn_id_len a parameter of the SSL_CTX or SSL_EVENT_CTX, so it will be the same across all the possible connections on the given BIO. IMO there is no point in varying the short_conn_id_len across connections on the same BIO.

t8m replied:
However the source host and port is necessary to determine the connection if we use 0 length connection ids.

paulidale replied:
The source host and port are necessary to know how long the connection ID will be.
Different QUIC implementations will have different lengths and we need to be able to distinguish between them.
Here, we are receiving packets from a different QUIC implementation.

Unless we do the splitting further down the stack.  Our diagram only has the Datagram BIO below the demuxer.  I don't think the splitting should be done there.

t8m replied:
The destination connection ID is our choice - it is equal to the source connection ID on the packets we send.

t8m replied:
And yes, the exception is the Initial packet from a client but that is a long header packet so short_conn_id_len does not apply.

openssl-machine commented:
This pull request is ready to merge

t8m commented:
Merged to master branch. Thank you for this great work!
